src/v/archival/adjacent_segment_merger.cc

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/adjacent_segment_merger.h"
+
+#include "archival/ntp_archiver_service.h"
+#include "archival/segment_reupload.h"
+#include "archival/types.h"
+#include "config/configuration.h"
+#include "model/fundamental.h"
+#include "storage/disk_log_impl.h"
+#include "utils/human.h"
+
+namespace archival {
+
+local_adjacent_segment_merger::local_adjacent_segment_merger(
+  ntp_archiver& parent, retry_chain_logger& ctxlog)
+  : _archiver(parent)
+  , _ctxlog(ctxlog)
+  , _segment_size(config::shard_local_cfg().log_segment_size.bind())
+  , _target_segment_size(
+      config::shard_local_cfg().cloud_storage_segment_size_target.bind())
+  , _min_segment_size(
+      config::shard_local_cfg().cloud_storage_segment_size_min.bind()) {}
+
+ss::future<> local_adjacent_segment_merger::stop() {
+    interrupt();
+    return _gate.close();
+}
+
+ss::future<> local_adjacent_segment_merger::run(retry_chain_node& rtc) {
+    ss::gate::holder h(_gate);
+    vlog(_ctxlog.debug, "Adjacent segment merger run begin");
+    auto scanner = [this](
+                     model::offset start_offset,
+                     const cloud_storage::partition_manifest& manifest)
+      -> std::optional<adjacent_segment_run> {
+        adjacent_segment_run run;
+        auto high_watermark = _target_segment_size().value_or(_segment_size());
```

Comment:
nit: naming: since we usually use "watermark" for offsets, maybe name these something different.  maybe they should be just called `target_segment_size` and `min_segment_size`, like the configs they're derived from.